### ISSUE TITLE
Add support for Realm Functions

### ISSUE THREAD

**Issue by cmelchior** [realm/realm-java#6807]
See [https://docs.google.com/document/d/1iD9ujY76BWVUOdN7rRH4XyPF9hHxx48RcGc-ZaQPyII/edit#heading=h.wt1zjxqe6taa](https://docs.google.com/document/d/1iD9ujY76BWVUOdN7rRH4XyPF9hHxx48RcGc-ZaQPyII/edit#heading=h.wt1zjxqe6taa)

Part of [https://github.com/realm/realm-java/issues/6801](https://github.com/realm/realm-java/issues/6801)

From PR #6810

* [x] Argument/result encoding/decoding support. Defaults to codec registry for primitive types, `BsonValue`s, and `List` and `Map` types
* [x] Backend support
* [x] Tests
* [x] Test of invoking function only authorized to specific user
* [x] Adobt final package/naming scheme
* [ ] Align async API across other MongoDB related APIs (will be done in separate task when set of API's are known)
* [ ] Ignored test due to C++ Parser issues (issue created and tracked in ROS-127)

**Comment by RealmBot** [comment]
➤ Brian Munkholm commented:

Currently blocked on OS work.

**Comment by RealmBot** [comment]
➤ Claus Rørbech commented:

OS Overall merge progress is tracked in ROBJSTORE-73
